enterprise/server/remote_execution/containers/ociruntime/ociruntime.go

```diff
@@ -1293,58 +1294,86 @@ func downloadLayer(ctx context.Context, layer ctr.Layer, destDir string) error {
+			return status.UnavailableErrorf("tar entry is not clean: %q", header.Name)
 		}
+		target := filepath.Join(tempUnpackDir, filepath.Clean(header.Name))
+		base := filepath.Base(target)
+		dir := filepath.Dir(target)
 
-		base := filepath.Base(path)
-		dir := filepath.Dir(path)
 		const whiteoutPrefix = ".wh."
-
-		// Directory whiteouts
-		if base == whiteoutPrefix+whiteoutPrefix+".opq" {
-			if err := unix.Setxattr(dir, "trusted.overlay.opaque", []byte{'y'}, 0); err != nil {
-				return fmt.Errorf("setxattr on deleted dir: %w", err)
-			}
-			if err := os.Remove(path); err != nil {
-				return fmt.Errorf("remove directory whiteout marker: %w", err)
+		// Handle whiteout
+		if strings.HasPrefix(base, whiteoutPrefix) {
+			// Directory whiteout
+			if base == whiteoutPrefix+whiteoutPrefix+".opq" {
+				if err := unix.Setxattr(dir, "trusted.overlay.opaque", []byte{'y'}, 0); err != nil {
+					return status.UnavailableErrorf("setxattr on deleted dir: %s", err)
+				}
+				continue
 			}
-			return nil
-		}
 
-		// File whiteouts
-		if strings.HasPrefix(base, whiteoutPrefix) {
+			// File whiteout: Mark the file for deletion in overlayfs.
 			originalBase := base[len(whiteoutPrefix):]
 			originalPath := filepath.Join(dir, originalBase)
 			if err := unix.Mknod(originalPath, unix.S_IFCHR, 0); err != nil {
-				return fmt.Errorf("mknod for whiteout marker: %w", err)
-			}
-			if err := os.Remove(path); err != nil {
-				return fmt.Errorf("remove directory whiteout marker: %w", err)
+				return status.UnavailableErrorf("mknod for whiteout marker: %s", err)
 			}
-			return nil
+			continue
 		}
 
-		return nil
-	})
-	if err != nil {
-		return status.UnavailableErrorf("walk layer dir: %s", err)
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if err := os.MkdirAll(target, os.FileMode(header.Mode)); err != nil {
+				return status.UnavailableErrorf("create directory: %s", err)
+			}
+		case tar.TypeReg:
+			if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+				return status.UnavailableErrorf("create directory: %s", err)
+			}
+			f, err := os.OpenFile(target, os.O_CREATE|os.O_WRONLY, os.FileMode(header.Mode))
+			if err != nil {
+				return status.UnavailableErrorf("create file: %s", err)
+			}
+			if _, err := io.Copy(f, tr); err != nil {
+				f.Close()
+				return status.UnavailableErrorf("copy file content: %s", err)
+			}
+			if err := f.Chown(header.Uid, header.Gid); err != nil {
+				f.Close()
+				return status.UnavailableErrorf("chown file: %s", err)
+			}
+			f.Close()
+		case tar.TypeSymlink:
+			if err := os.Symlink(header.Linkname, target); err != nil {
+				return status.UnavailableErrorf("create symlink: %s", err)
+			}
+			if err := os.Lchown(target, header.Uid, header.Gid); err != nil {
+				return status.UnavailableErrorf("chown link: %s", err)
+			}
+		case tar.TypeLink:
+			if slices.Contains(strings.Split(header.Linkname, string(os.PathSeparator)), "..") {
```

Comment:
Could you add a comment to explain why we do this for hardlinks but not symlinks?

Reply:
Will do.

Reply:
As I work on the hardlink change, I also realize that symlinks are not only evaled inside the container, if a hardlink is targeting the symlink, which points to a host-level file, it will result in a breakout 🤔 

Let me study how podman are docker are doing this and mirror their approach.

Reply:
When creating a hardlink to a symlink, the symlink shouldn't be followed, so I don't think this is an issue? Creating a hardlink to a symlink should just create that same symlink at a new path (the hardlink destination), not the file it points to.